### ISSUE TITLE
Compatibility errors when building Terra on Ubuntu 15.10 (g++ 5.2)

### ISSUE THREAD

**Issue by paniq** [zdevito/terra#131]
Building Terra on Ubuntu 15.10 with llvm 3.6.2 fails with several linker errors due to missing symbols. 

g++ reports version 5.2.1-22ubuntu2 (although iirc g++ isn't even used here?), g++ was 4.9 in 15.04. It appears clang++ uses system headers during compilation.

I'm building with `ENABLE_CUDA = 0`

```
/home/lritter/devel/duangle/liminal/build/bin/clang++ -o /dev/null build/tkind.o build/tcompiler.o build/tllvmutil.o build/tcwrapper.o build/tinline.o build/terra.o build/lparser.o build/lstring.o build/lobject.o build/lzio.o build/llex.o build/lctype.o build/treadnumber.o build/tcuda.o build/tdebug.o build/tinternalizedfiles.o build/main.o build/linenoise.o build/LuaJIT-2.0.4/src/libluajit.a -L/home/lritter/devel/duangle/liminal/build/lib  -L/home/lritter/devel/duangle/liminal/build/lib -lclangFrontend -lclangDriver -lclangSerialization -lclangCodeGen -lclangParse -lclangSema -lclangAnalysis -lclangEdit -lclangAST -lclangLex -lclangBasic -lLLVMLTO -lLLVMObjCARCOpts -lLLVMLinker -lLLVMBitWriter -lLLVMIRReader -lLLVMAsmParser -lLLVMR600CodeGen -lLLVMipo -lLLVMVectorize -lLLVMR600AsmParser -lLLVMR600Desc -lLLVMR600Info -lLLVMR600AsmPrinter -lLLVMSystemZDisassembler -lLLVMSystemZCodeGen -lLLVMSystemZAsmParser -lLLVMSystemZDesc -lLLVMSystemZInfo -lLLVMSystemZAsmPrinter -lLLVMHexagonDisassembler -lLLVMHexagonCodeGen -lLLVMHexagonDesc -lLLVMHexagonInfo -lLLVMNVPTXCodeGen -lLLVMNVPTXDesc -lLLVMNVPTXInfo -lLLVMNVPTXAsmPrinter -lLLVMCppBackendCodeGen -lLLVMCppBackendInfo -lLLVMMSP430CodeGen -lLLVMMSP430Desc -lLLVMMSP430Info -lLLVMMSP430AsmPrinter -lLLVMXCoreDisassembler -lLLVMXCoreCodeGen -lLLVMXCoreDesc -lLLVMXCoreInfo -lLLVMXCoreAsmPrinter -lLLVMMipsDisassembler -lLLVMMipsCodeGen -lLLVMMipsAsmParser -lLLVMMipsDesc -lLLVMMipsInfo -lLLVMMipsAsmPrinter -lLLVMAArch64Disassembler -lLLVMAArch64CodeGen -lLLVMAArch64AsmParser -lLLVMAArch64Desc -lLLVMAArch64Info -lLLVMAArch64AsmPrinter -lLLVMAArch64Utils -lLLVMARMDisassembler -lLLVMARMCodeGen -lLLVMARMAsmParser -lLLVMARMDesc -lLLVMARMInfo -lLLVMARMAsmPrinter -lLLVMPowerPCDisassembler -lLLVMPowerPCCodeGen -lLLVMPowerPCAsmParser -lLLVMPowerPCDesc -lLLVMPowerPCInfo -lLLVMPowerPCAsmPrinter -lLLVMSparcDisassembler -lLLVMSparcCodeGen -lLLVMSparcAsmParser -lLLVMSparcDesc -lLLVMSparcInfo -lLLVMSparcAsmPrinter -lLLVMTableGen -lLLVMDebugInfo -lLLVMOption -lLLVMX86Disassembler -lLLVMX86AsmParser -lLLVMX86CodeGen -lLLVMSelectionDAG -lLLVMAsmPrinter -lLLVMX86Desc -lLLVMMCDisassembler -lLLVMX86Info -lLLVMX86AsmPrinter -lLLVMX86Utils -lLLVMMCJIT -lLLVMLineEditor -lLLVMInstrumentation -lLLVMInterpreter -lLLVMExecutionEngine -lLLVMRuntimeDyld -lLLVMCodeGen -lLLVMScalarOpts -lLLVMProfileData -lLLVMObject -lLLVMMCParser -lLLVMBitReader -lLLVMInstCombine -lLLVMTransformUtils -lLLVMipa -lLLVMAnalysis -lLLVMTarget -lLLVMMC -lLLVMCore -lLLVMSupport -ldl -pthread -g -Wl,-t | egrep "lib(LLVM|clang)"  > build/dep_objects/llvm_list
build/tcompiler.o: In function `terra_inittarget(lua_State*)':
/home/lritter/devel/duangle/liminal/build/src/terra/src/tcompiler.cpp:290: undefined reference to `llvm::TargetRegistry::lookupTarget(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)'
build/tcompiler.o: In function `llvm::sys::DynamicLibrary::LoadLibraryPermanently(char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
/home/lritter/devel/duangle/liminal/build/include/llvm/Support/DynamicLibrary.h:78: undefined reference to `llvm::sys::DynamicLibrary::getPermanentLibrary(char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)'
build/tcwrapper.o: In function `dofile(terra_State*, TerraTarget*, char const*, char const**, char const**, Obj*)':
/home/lritter/devel/duangle/liminal/build/src/terra/src/tcwrapper.cpp:849: undefined reference to `clang::CreateLLVMCodeGen(clang::DiagnosticsEngine&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, clang::CodeGenOptions const&, clang::TargetOptions const&, llvm::LLVMContext&, clang::CoverageSourceInfo*)'
build/tllvmutil.o: In function `llvmutil_disassemblefunction':
/home/lritter/devel/duangle/liminal/build/src/terra/src/tllvmutil.cpp:97: undefined reference to `llvm::TargetRegistry::lookupTarget(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)'
build/tllvmutil.o: In function `llvmutil_executeandwait(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, char const**, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
/home/lritter/devel/duangle/liminal/build/src/terra/src/tllvmutil.cpp:348: undefined reference to `llvm::sys::ExecuteNoWait(llvm::StringRef, char const**, char const**, llvm::StringRef const**, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*)'
/usr/bin/ld: link errors found, deleting executable `/dev/null'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
# extract needed LLVM objects based on a dummy linker invocation
< build/dep_objects/llvm_list build/LuaJIT-2.0.4/src/luajit src/unpacklibraries.lua build/dep_objects
# include all luajit objects, since the entire lua interface is used in terra 
cd build/dep_objects/luajit; ar x ../../../build/LuaJIT-2.0.4/src/libluajit.a
mkdir -p release/lib
rm -f release/lib/libterra.a
ar -cq release/lib/libterra.a build/tkind.o build/tcompiler.o build/tllvmutil.o build/tcwrapper.o build/tinline.o build/terra.o build/lparser.o build/lstring.o build/lobject.o build/lzio.o build/llex.o build/lctype.o build/treadnumber.o build/tcuda.o build/tdebug.o build/tinternalizedfiles.o build/dep_objects/*/*.o
ranlib release/lib/libterra.a
mkdir -p release/bin release/lib
/home/lritter/devel/duangle/liminal/build/bin/clang++ -shared -fPIC -Wl,-export-dynamic -Wl,--whole-archive release/lib/libterra.a -Wl,--no-whole-archive -ldl -pthread -o release/lib/libterra_dynamic.so  
/home/lritter/devel/duangle/liminal/build/bin/clang++ build/main.o build/linenoise.o -o release/bin/terra -g -Wl,-export-dynamic -Wl,--whole-archive release/lib/libterra.a -Wl,--no-whole-archive  -ldl -pthread
release/lib/libterra.a(tcompiler.o): In function `terra_inittarget(lua_State*)':
/home/lritter/devel/duangle/liminal/build/src/terra/src/tcompiler.cpp:290: undefined reference to `llvm::TargetRegistry::lookupTarget(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)'
release/lib/libterra.a(tcompiler.o): In function `llvm::sys::DynamicLibrary::LoadLibraryPermanently(char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
/home/lritter/devel/duangle/liminal/build/include/llvm/Support/DynamicLibrary.h:78: undefined reference to `llvm::sys::DynamicLibrary::getPermanentLibrary(char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)'
release/lib/libterra.a(tcwrapper.o): In function `dofile(terra_State*, TerraTarget*, char const*, char const**, char const**, Obj*)':
/home/lritter/devel/duangle/liminal/build/src/terra/src/tcwrapper.cpp:849: undefined reference to `clang::CreateLLVMCodeGen(clang::DiagnosticsEngine&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, clang::CodeGenOptions const&, clang::TargetOptions const&, llvm::LLVMContext&, clang::CoverageSourceInfo*)'
release/lib/libterra.a(tllvmutil.o): In function `llvmutil_disassemblefunction':
/home/lritter/devel/duangle/liminal/build/src/terra/src/tllvmutil.cpp:97: undefined reference to `llvm::TargetRegistry::lookupTarget(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&)'
release/lib/libterra.a(tllvmutil.o): In function `llvmutil_executeandwait(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, char const**, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*)':
/home/lritter/devel/duangle/liminal/build/src/terra/src/tllvmutil.cpp:348: undefined reference to `llvm::sys::ExecuteNoWait(llvm::StringRef, char const**, char const**, llvm::StringRef const**, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >*, bool*)'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Makefile:182: recipe for target 'release/bin/terra' failed
```

**Comment by paniq** [comment]
Found a fix.

According to https://github.com/JuliaLang/julia/issues/10201, the problem was brought about by ``-D_GLIBCXX_USE_CXX11_ABI=1`` being default since g++ 5.1. 

I changed Terra's Makefile to edit line 64 to
```
ifneq ($(LLVM_VERSION), 32)
CPPFLAGS = -std=c++11 -Wno-c++11-narrowing -D_GLIBCXX_USE_CXX11_ABI=0
endif
```

and that fixes the problem for with the official LLVM binaries for 3.6.2. However it would create new problems for users who built their LLVM from source using g++ 5.1.

It would be useful to be able to change CPPFLAGS from the command line before building, so people using prebuilt binaries like myself could inject a ``-D_GLIBCXX_USE_CXX11_ABI=0`` option. I didn't see a way to do it with the current Makefile.
